### PR TITLE
Partial source_cache invalidation from geojson_source

### PR DESCRIPTION
**Pull request by mfedderly**
Follow-on work to #1236 which lets us skip recomputing tiles that we know have not changed.

Introduce TileBitmask which keeps track of the tiles that actually changed.
Consume TileBitmask in the geojson_worker_source to keep track of the changes we send through in a diff.
Pass this back to the main thread into source_cache to selectively invalidate tiles instead of hitting all of them.

How well this works depends on the usage patterns. For the geojson-update debug page it looked like we were able to skip invalidating individual tiles about 40% of the time at a zoomed out view, but about 85% of the time when we zoom in on a smaller area. The geographic distribution of updates may also impact how often we can skip invalidating our tiles.

Once we're happy with the approach I can go ahead and polish the rest of this with tests, changelog, etc.

image

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Manually test the debug page.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/source/geojson_source_diff.ts [hunk 1]
```diff
@@ -130,6 +144,105 @@ export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Featur
...
                 }
             }
+
+            // invalidate the updated feature
+            markInvalidated(invalidated, feature);
+        }
+    }
+}
+
+function markInvalidated(invalidated: TileBitmask, feature: GeoJSON.Feature | undefined) {
```
Comment (mfedderly) [inline src/source/geojson_source_diff.ts:154]: I've considered that this might be better as part of TileBitmask itself.

File: src/source/geojson_source_diff.ts [hunk 2]
```diff
@@ -130,6 +144,105 @@ export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Featur
...
+    if (result < 0) {
+        return 0;
+    } else if (result > TileBitmask.MaxZoomTiles) {
+        return TileBitmask.MaxZoomTiles;
+    } else {
+        return result;
+    }
+}
+
+function getBbox(geometry: GeoJSON.Geometry, bbox: [number, number, number, number] = [Infinity, Infinity, -Infinity, -Infinity]): [number, number, number, number] {
```
Comment (mfedderly) [inline src/source/geojson_source_diff.ts:205]: This bbox logic is pretty similar to some other parts of the code. I'm not sure whether we should try to combine this with the existing code or not. it also exists almost verbatim in @turf/bbox but I figured we probably didn't want to add a whole dependency there.

File: src/util/tile_bitmask.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        y: coord & mask,
+    };
+}
+
+enum Composition {
+    MIXED = 0,
+    MARKED = 1,
+}
+
+export type SerializedTileBitmask = Int32Array;
```
Comment (mfedderly) [inline src/util/tile_bitmask.ts:23]: @westinrm I don't fully understand how this is used, is Int32Array correct? Or can we maybe get away with Uint16Array?

File: src/source/geojson_source_diff.ts [hunk 4]
```diff
@@ -130,6 +144,105 @@ export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Featur
...
+        } else {
+            // create a bbox from scratch
+            bbox = getBbox(feature.geometry);
+        }
+
+        // store off the bbox for reuse later
+        feature.bbox = bbox;
+    }
+
+    const [minX, minY, maxX, maxY] = bbox;
```
Comment (mfedderly) [inline src/source/geojson_source_diff.ts:173]: @westinrm What happens when the min/max X values are outside of [-180, 180] here? The longitudes can potentially be outside that range when we want to specify features that cross the antimeridian. I suspect we need to fix something here.

File: src/source/geojson_source_diff.ts [hunk 5]
```diff
@@ -112,6 +124,8 @@ export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Featur
             }
 
             if (update.newGeometry) {
+                // need to clear the bbox when we change the geometry, it'll be recalculated on demand
```
Comment (HarelM) [inline src/source/geojson_source_diff.ts:127]: This is not related to this change, right? It's a different bug fix...?

Comment (mfedderly) [reply src/source/geojson_source_diff.ts:127]: Before this we didn't consume the bbox on the worker thread so it didn't matter. Unless you can get it back out using `map.queryRenderedFeatures()` I'm not sure anyone even had access to it.

**Comment by HarelM** [comment]
For the sake of clarity can you attach a flow diagram of the data being passed between the classes?
Is there an alternative way to do it? The current back and forth is a bit too complicated and I'm hoping there's an easier way?
Also wouldn't it be more efficient to focus on the geojson index part of things? I feel like we are walking around it instead of fixing it...

File: src/source/geojson_source_diff.ts (hunk 5 above)
Comment (HarelM) [reply src/source/geojson_source_diff.ts:127]: Ah, OK thanks.

**Comment by mfedderly** [comment]
The flow of data winds up being a single direction:

applySourceDiff now mutates the TileBitmask as it performs its update
The TileBitmask is then returned from GeoJSONWorkerSource to GeoJSONSource (in serialized form)
GeoJSONSource adds it to the 'content' data event
source_cache will perform partial tile invalidation for the 'content' data event if it has a TileBitmask

This concept can also be useful on other source types as well. It would be interesting to add handling for `304 Not Modified` to the vector source.

One thing that might make it safer to handle Evented and worker.postMessage would be to use something like [strict-event-emitter](https://www.npmjs.com/package/strict-event-emitter) which gives you type safety around these events. I also have a version that works with worker messaging as well that I could open source if you were interested. This can be done entirely in typescript without any javascript.

With regard to the geojson index comment, all of the changes so far have been to improve CPU usage on the main JS thread. The main thread time is the most important resource to guard since it is shared with all of the other code on the page. These changes will not be irrelevant if we clean up memory/cpu usage on the worker thread, they will be complementary.

**Comment by HarelM** [comment]
Sorry, my request might not have been clear. I would like to see a flow diagram using visuals, something like the following, preferably at the initial message of this PR, so that if someone will review this later on, s.he will be able to see the idea behind this PR more easily.
image
Geojson source worker memory usage is also relevant as we have seen some bad behavior on mobile browsers that have limited resources. 
I would like to make sure we address this part as soon as possible since current situation is less than optimal in terms of memory (when looking at large geojsons)...

In terms of event safety, yes, I think this needs a lot more attention and care as the current flow of events is brittle. But this should not be part of this PR, obvuisly.

**Comment by HarelM** [comment]
Draw.io is great.

**Comment by mfedderly** [comment]
Can't upload the draw.io file itself but here's a screenshot:

image

File: src/source/geojson_source.ts [hunk 6]
```diff
@@ -294,7 +295,7 @@ class GeoJSONSource extends Evented implements Source {
                 return;
             }
 
-            const data: any = {dataType: 'source'};
+            const data: any = {dataType: 'source', invalidated: result?.invalidated ? TileBitmask.deserialize(result.invalidated) : undefined};
```
Comment (HarelM) [inline src/source/geojson_source.ts:298]: I would prefer to keep the events data as pure JSON objects and not introduce some classes that are hard to use. I would prefer to do the deserialization only where the data is used.

File: src/source/geojson_source_diff.ts (hunk 2 above)
Comment (HarelM) [reply src/source/geojson_source_diff.ts:206]: If you can export it to a file and use the function in multiple places it's better than to duplicate.

File: src/source/geojson_worker_source.ts [hunk 7]
```diff
@@ -242,7 +249,9 @@ class GeoJSONWorkerSource extends VectorTileWorkerSource {
             }
         } else if (params.dataDiff) {
             if (this._dataUpdateable) {
-                applySourceDiff(this._dataUpdateable, params.dataDiff, promoteId);
+                const invalidated = new TileBitmask();
```
Comment (HarelM) [inline src/source/geojson_worker_source.ts:252]: Can we return `invalidated` instead of adding this as a parameter which we change?

File: src/ui/events.ts [hunk 8]
```diff
@@ -48,6 +49,7 @@ export interface MapSourceDataEvent extends MapLibreEvent {
     sourceId: string;
     sourceDataType: MapSourceDataType;
     tile: any;
+    invalidated?: TileBitmask;
```
Comment (HarelM) [inline src/ui/events.ts:52]: Can we use the serialized version here?

File: src/util/tile_bitmask.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+/**
+ * Stores a mask of tile coordinates as a quadtree.
+ */
+export class TileBitmask {
```
Comment (HarelM) [inline src/util/tile_bitmask.ts:19]: This class needs to be fully covered with tests.
Can we use kdbush here or this is completely unrelated?
Do we really need to write all this? isn't there somewhere in the code something similar?

File: src/util/tile_bitmask.ts (hunk 3 above)
Comment (westinrm) [reply src/util/tile_bitmask.ts:14]: `packTileCoord(...) << 2` produces `zoom << 16`.  So it can only be a UInt16 if you reduce the max tree depth.
Should be possible to be unsigned though.

File: src/util/tile_bitmask.ts (hunk 9 above)
Comment (westinrm) [reply src/util/tile_bitmask.ts:19]: This isn't a spatial index, it's a compressed bitmask that acts as a very compact `set<TileCoordinate>`.  So kdbush wouldn't work.
I don't know of any existing places in maplibre or any NPM packages that have the same code.

File: src/source/geojson_source_diff.ts (hunk 4 above)
Comment (westinrm) [reply src/source/geojson_source_diff.ts:174]: Probably should just make `convertToTileCoord` clamp to the tile range.

File: src/source/geojson_source_diff.ts [hunk 10]
```diff
@@ -130,6 +144,106 @@ export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Featur
...
+        }
+    }
+}
+
+function markInvalidated(updateable: Map<GeoJSONFeatureId, GeoJSON.Feature>, id: GeoJSONFeatureId, invalidated: TileBitmask, feature: GeoJSON.Feature | undefined) {
+    if (!feature || !feature.geometry) {
+        return;
+    }
+
+    let bbox = feature.bbox;
```
Comment (westinrm) [inline src/source/geojson_source_diff.ts:159]: This is the lat/long bounding box, `convertToTileCoord` expects normalized web mercator coordinates.
Internally this code accepts WorkerFeature, which has the geometry after https://github.com/mapbox/geojson-vt/blob/main/src/convert.js (reprojection) happens.

**Comment by mfedderly** [comment]
Did some further testing with this and I don't think the performance gains justify the complexity here.
